runtime/src/accounts_db.rs

```diff
@@ -921,6 +949,96 @@ impl AccountsDB {
         reclaims
     }
 
+    fn squash_uncleaned_keys(&self, max_slot: Slot) {
+        let (keys, mut slots) = self.collect_uncleaned_keys_to_slot(max_slot);
+        if slots.is_empty() {
+            return;
+        }
+
+        slots.sort_unstable();
+        let max_slot = *slots.last().unwrap();
+        let mut squashed_keys = self.squashed_uncleaned_keys.write().unwrap();
+        for key in keys.into_iter().flatten() {
+            squashed_keys.0.insert(key);
+        }
+        squashed_keys.1 = max_slot;
+    }
+
+    fn collect_uncleaned_keys_to_slot(&self, max_slot: Slot) -> (Vec<Vec<Pubkey>>, Vec<Slot>) {
+        let slots: Vec<Slot> = self
+            .uncleaned_keys
+            .iter()
+            .filter_map(|entry| {
+                let slot = entry.key();
+                if *slot <= max_slot {
+                    Some(*slot)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        (
+            slots
+                .iter()
+                .filter_map(|slot| {
+                    let keys = self.uncleaned_keys.remove(slot).unwrap().1;
+                    if self.accounts_index.is_root(*slot) {
+                        Some(keys)
+                    } else {
+                        None
+                    }
+                })
+                .collect(),
+            slots,
+        )
+    }
+
+    // Construct a vec of the keys needed for cleaning from uncleaned_keys and
+    // zero_lamport_keys sets.
+    fn get_clean_keys(
+        &self,
+        max_clean_root: Option<Slot>,
+        timings: &mut CleanKeyTimings,
+    ) -> Vec<Pubkey> {
+        let mut zero_lamport_key = Measure::start("zero_lamport_key");
+        let mut pubkeys = self.accounts_index.zero_lamport_accounts.clone();
+        timings.zero_lamport_count = pubkeys.len() as u64;
+        zero_lamport_key.stop();
+        timings.zero_lamport_key += zero_lamport_key.as_us();
+
+        let mut key_create = Measure::start("key_create");
+        let max_slot = max_clean_root.unwrap_or_else(|| self.accounts_index.max_root());
+        let delta_keys: Vec<_> = { self.collect_uncleaned_keys_to_slot(max_slot).0 };
+        key_create.stop();
+        timings.key_create += key_create.as_us();
+
+        self.thread_pool_clean.install(|| {
+            delta_keys.par_iter().for_each(|keys| {
+                for key in keys {
+                    pubkeys.insert(*key);
+                }
+            });
+        });
+
+        {
+            let mut squashed_uncleaned_keys = self.squashed_uncleaned_keys.write().unwrap();
```

Comment:
nits: `let (mut squashed_uncleaned_keys, squashed_root) = ...`

Reply:
same as this: https://github.com/solana-labs/solana/pull/14652#discussion_r562024379